USB: Add train controller emulation

joestringer opened the pull request:
### Description of Changes

Emulate a range of "Densha Mascon" train controllers originally sold by Taito, such as TCP20009, TCP20011, TCP20014. Built using datasheets courtesy of @marcriera and other contributors at https://marcriera.github.io/ddgo-controller-docs/ .

Type 2 | Shinkansen | Ryojōhen
:-----:|:----------:|:--------:
![type2](https://github.com/user-attachments/assets/3d1999a1-f8a0-4c81-99e8-4a7453d9c327) | ![shinkansen](https://github.com/user-attachments/assets/198ce632-6742-4e6a-be82-2933f98b3a31) | ![ryojouhen](https://github.com/user-attachments/assets/210a7481-2173-49f6-85c9-da193a507e3e)

### Rationale behind Changes

Lever-based controllers provide an advantage over the standard DualShock emulation for some games, as these controllers physically hold an axis in a particular level without requiring the player to actively apply pressure. These changes allow players with lever-based controllers to play Densha de Go! titles such as:

- Densha de GO! 3 Tsūkin-hen
- Densha de GO! Final
- Densha de GO! Professional 2
- Densha de GO! Ryojōhen
- Densha de GO! San'yō Shinkansen-hen

### Suggested Testing Steps

The [ddgo-controller-docs website](https://marcriera.github.io/ddgo-controller-docs/) describes these controllers and lists the compatible titles. I have tested these with my own USB Zuiki One-Handle MasCon controller for Nintendo Switch. Given the way that the hardware maps into the emulated controller, the best subtype for this particular hardware controller is the "Type 2" controller. Some games will change the input scheme and number of power/brake notch settings based on the selected subtype variant.

Related: #4763

joestringer commented:
cc @sonik-br this might be interesting to you. I have not tested the "USB passthrough" mode but I believe that this should allow original Densha De GO! USB controllers to work correctly with this patch.

joestringer commented:
Fixed a couple of CI complaints - extraneous xml tags in the vcxproj file, and I forgot to add the new files to the CMakeLists.txt, leading to linker failures in the CI builds. These should be resolved now.

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+{
+	const char* MasconDevice::Name() const
+	{
+		return TRANSLATE_NOOP("USB", "Densha Controller");
```
```suggestion
		return TRANSLATE_NOOP("USB", "Train Controller");
```

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+			case MT_SHINKANSEN:
+			{
+				static constexpr const InputBindingInfo bindings[] = {
+					{"Power", TRANSLATE_NOOP("Pad", "Power"), ICON_PF_LEFT_ANALOG_DOWN, InputBindingInfo::Type::Axis, CID_MC_POWER, GenericInputBinding::LeftStickDown},
```
```suggestion
					{"Power", TRANSLATE_NOOP("USB", "Power"), ICON_PF_LEFT_ANALOG_DOWN, InputBindingInfo::Type::Axis, CID_MC_POWER, GenericInputBinding::LeftStickDown},
```
same for the entire file

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+		return (get_ab(buttons) | (button_at(buttons, CID_MC_CAMERA) >> 4) | ((get_cd(buttons) | get_ss(buttons)) << 1));
+	}
+
+	int MasconState::TokenIn(u8* buf, int len)
```
These TokenIn methods are left over from the time when there were multiple variants, one for each API, now the body can be moved directly in mascon_handle_data()

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.h:
```diff
@@ -0,0 +1,240 @@
+
+	DEFINE_DCT_DEV_DESCRIPTOR(dct01, 0x04, 0x0004);
+
+	// https://marcriera.github.io/ddgo-controller-docs/controllers/usb/descriptors/tcpp20009_hid-report-descriptor.txt
+	static const uint8_t dct01_hid_report_descriptor[] = {
+		MASCON_PREAMBLE,
+		MASCON_PAD_BYTE,
+		MASCON_AXES, // Brake, Power, Horn
+		MASCON_DPAD,
+		MASCON_BUTTONS(6),
+		MASCON_END,
+	};
```
`*_hid_report_descriptor` and `#define MASCON_*` are unused. These controllers aren't HID class devices, but Vendor specific : bDeviceClass=0xff.

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+		return nullptr;
+	}
+
+	float MasconDevice::GetBindingValue(const USBDevice* dev, u32 bind_index) const
+	{
+		const MasconState* s = USB_CONTAINER_OF(dev, const MasconState, dev);
+		return s->GetBindValue(bind_index);
+	}
+
+	void MasconDevice::SetBindingValue(USBDevice* dev, u32 bind_index, float value) const
+	{
+		MasconState* s = USB_CONTAINER_OF(dev, MasconState, dev);
+		s->SetBindValue(bind_index, value);
+	}
```
GetBindingValue/GetBindValue and SetBindingValue/SetBindValue can be merged.

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+// SPDX-FileCopyrightText: 2002-2024 PCSX2 Dev Team
+// SPDX-License-Identifier: GPL-3.0+
+
+#include "usb-mascon.h"
```
Please rename mascon (`usb-mascon.cpp / usb-mascon.h / DEVTYPE_MASCON / MasconDevice / MasconState`) to something suitable for both DDG and TS controllers like usb-train.cpp, DEVTYPE_TRAIN, etc

Florin9doi commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+					{"Down", TRANSLATE_NOOP("Pad", "D-Pad Down"), ICON_PF_DPAD_DOWN, InputBindingInfo::Type::Button, CID_MC_DOWN, GenericInputBinding::DPadDown},
+					{"Left", TRANSLATE_NOOP("Pad", "D-Pad Left"), ICON_PF_DPAD_LEFT, InputBindingInfo::Type::Button, CID_MC_LEFT, GenericInputBinding::DPadLeft},
+					{"Right", TRANSLATE_NOOP("Pad", "D-Pad Right"), ICON_PF_DPAD_RIGHT, InputBindingInfo::Type::Button, CID_MC_RIGHT, GenericInputBinding::DPadRight},
+					{"A", TRANSLATE_NOOP("Pad", "A Button"), ICON_PF_KEY_A, InputBindingInfo::Type::Button, CID_MC_A, GenericInputBinding::Square},
```
```suggestion
					{"A", TRANSLATE_NOOP("Pad", "A"), ICON_PF_KEY_A, InputBindingInfo::Type::Button, CID_MC_A, GenericInputBinding::Square},
```

IlDucci commented on pcsx2/USB/usb-pad/usb-mascon.cpp:
```diff
@@ -0,0 +1,585 @@
+		static const char* subtypes[] = {
+			TRANSLATE_NOOP("USB", "Type 2"),
+			TRANSLATE_NOOP("USB", "Shinkansen"),
+			TRANSLATE_NOOP("USB", "Ryojouhen"),
```
Typo here? In other areas this is read as "Ryojōhen".

joestringer replied:
I wasn't sure whether unicode characters were allowed in source :) Fixed.

IlDucci commented on pcsx2-qt/Settings/USBBindingWidget_DenshaCon.ui:
```diff
@@ -0,0 +1,755 @@
+        <item>
+         <widget class="QLabel" name="label_2">
+          <property name="text">
+           <string>The Densha Controller Type 2 is a two-handle train master controller with a D-Pad and six buttons. The power axis starts from neutral and has five progressively higher notches of power. The brake axis starts released, and has eight progressively increasing notches plus an Emergency Brake. This controller is compatible with several games and generally maps well to hardware with  ...
```
Could you please add here, as well as in the other two large descriptions, an extracomment="COMMENT" inside the `<string>` tag for localization purposes? I would like to have a description of the term" mascon" and to point out that there are no official translations for the controller name.

IlDucci commented on pcsx2-qt/Settings/USBBindingWidget_DenshaCon.ui:
```diff
@@ -0,0 +1,755 @@
+       <item>
+        <widget class="QGroupBox" name="groupBox_32">
+         <property name="title">
+          <string>Power Lever</string>
```
Could you please add here, as well as in the other copies of this string on UI files, an extracomment="COMMENT" inside the `<string>` tag for localization purposes? I would like to describe "Power Lever", as in, if "Power" means "Accelerator".

IlDucci commented on pcsx2-qt/Settings/USBBindingWidget_RyojouhenCon.ui:
```diff
@@ -0,0 +1,762 @@
+        <item>
+         <widget class="QLabel" name="label_2">
+          <property name="text">
+           <string>The Ryojōhen Controller is a two-handle train master controller with a D-Pad and seven buttons. The power axis starts from neutral and has four progressively higher notches of power. The brake axis is analogue with three regions representing to release brake pressure, hold brakes at the current pressure, or increase brake pressure. Extending this axis to the extreme triggers th ...
```
Could you please add here, as well as in the other copies of this string on UI files, an extracomment="COMMENT" inside the `<string>` tag for localization purposes? I would like to mention all translators here to keep the romanization of Ryojōhen consistent across the entire translation for Western languages, as well as any notes you might want to add for people who have no idea what a Densha de Go controller does.

IlDucci commented on pcsx2-qt/Settings/USBBindingWidget_RyojouhenCon.ui:
```diff
@@ -0,0 +1,762 @@
+        <item row="1" column="2">
+         <widget class="QGroupBox" name="groupBox_20">
+          <property name="title">
+           <string>Announce</string>
```
Could you please add here, as well as in the other copies of this string on UI files, an extracomment="COMMENT" inside the `<string>` tag for localization purposes? I would like to explain all translators what are the announcements that can be done with this button.

IlDucci reviewed (commented):
Here's some suggestions to make life a bit easier for translators.

sonik-br commented:
> cc @sonik-br this might be interesting to you. I have not tested the "USB passthrough" mode but I believe that this should allow original Densha De GO! USB controllers to work correctly with this patch.

Amazing work!
I can try and report back. There's a build to download or I need to build from source?

kamfretoz commented:
> There's a build to download or I need to build from source?

You can download the CI artifacts from the checks tab, make sure to pick your OS.

![image](https://github.com/user-attachments/assets/290cb115-43b5-487e-9391-fe28defef6a8)

sonik-br commented:
Not sure how to use usb passthrough. The device is not HID and does not shows as a gamepad under windows.
I need to install generic a usb driver with zadig?

Florin9doi commented:
> Not sure how to use usb passthrough. The device is not HID and does not shows as a gamepad under windows. I need to install generic a usb driver with zadig?

Won't work. Only HID controllers like DGOC-44U will work

sonik-br commented:
> Won't work. Only HID controllers like DGOC-44U will work

Ah ok!
I can emulate this device using a rp2040 then. Can try it later this week.

joestringer commented:
@Florin9doi @IlDucci thanks for the reviews! I believe I've addressed all of your comments.

Florin9doi commented on pcsx2/USB/usb-pad/usb-train.cpp:
```diff
@@ -0,0 +1,546 @@
+				return;
+		}
+
+		usb_packet_copy(p, &out, sizeof(out));
```
Although wMaxPacketSize is equal to 8 bytes for all 3 controllers, 2 of them send only 6 bytes (?): 
https://marcriera.github.io/ddgo-controller-docs/controllers/usb/tcpp20009/#input
https://marcriera.github.io/ddgo-controller-docs/controllers/usb/tcpp20011/#input

I'd split TrainConOut in 3 distinct structs

joestringer replied:
Makes sense. I found that even if sending 8 bytes, the other controllers seemed to work fine. But your suggestion is cleaner, I'll fix that up.

Florin9doi commented on pcsx2/pcsx2.vcxproj.filters:
```diff
@@ -2413,4 +2419,4 @@
   </ItemGroup>
-</Project>
\ No newline at end of file
+</Project>
```
newline

Florin9doi commented on pcsx2/USB/usb-pad/usb-train.cpp:
```diff
@@ -0,0 +1,546 @@
+		return nullptr;
+	}
+
+} // namespace usb_pad
```
newline

Florin9doi replied:
the newline is missing again

joestringer replied:
There's something about my Visual Studio configuration that seems to leak this back in. It's probably a line feed setting or something, but I don't usually use this dev environment so the settings are a bit non-obvious to me. I'll take a look at this.

IlDucci commented on pcsx2-qt/Settings/USBBindingWidget_RyojouhenCon.ui:
```diff
@@ -0,0 +1,762 @@
+        </size>
+       </property>
+       <property name="title">
+        <string>Door Buttons</string>
```
Please add a comment for translators so they know this a button to open the train's doors.

IlDucci reviewed (commented):
Thanks for hearing translators out! Unfortunately, I missed one element that could probably use a hint for translators (who might not expect train stuff happening in an emulator).

sonik-br commented:
I could no get passthrough to work.
I'm emulating a PC Two Handle (DGOC-44U) using the descriptors from marcriera.
Need any special setting?

Florin9doi commented:
What kind of passthrough do you expect? You'll still have to map the axes and buttons , but the axes values won't stick to predefined values and will be sent unchanged.

sonik-br commented:
> What kind of passthrough do you expect? You'll still have to map the axes and buttons , but the axes values won't stick to predefined values and will be sent unchanged.

Brake axis can't be mapped. It does not pick any movement on the map screen.
But the axis is working as I can test on windows and on the PC version of the game (DDGO Final)

sonik-br commented:
Still not working.
Tested with fullaxis0 and fullaxis1.

```
[USB1]
Type = DenshaCon
DenshaCon_A = DInput-0/Button1
DenshaCon_B = DInput-0/Button0
DenshaCon_C = DInput-0/Button2
DenshaCon_D = DInput-0/Button3
DenshaCon_Select = DInput-0/Button4
DenshaCon_Start = DInput-0/Button5
DenshaCon_Up = DInput-0/Button6
DenshaCon_Right = DInput-0/Button9
DenshaCon_Down = DInput-0/Button7
DenshaCon_Left = DInput-0/Button8
DenshaCon_Power = DInput-0/+Axis1~
DenshaCon_Brake = DInput-0/FullAxis1
DenshaCon_Passthrough = true
```

joestringer commented:
@Florin9doi @IlDucci I've resolved your latest feedback 🚀

@sonik-br thanks for testing. Do the other buttons and power lever work in-game or are all of the controls broken? If they work, does the lever increase the number of power notches that you expect?

sonik-br commented:
@joestringer buttons and dpad works in-game.
Power lever does not. The game think it's always on max power.
![image](https://github.com/user-attachments/assets/4340a6a3-aeda-433f-97fe-3655f9508331)

joestringer commented:
@sonik-br Hmm OK. Early on I did encounter similar issues while initially developing the support. I can't do much about the brake since that relies on the input library to pull the values into the point where this feature processes it. Unless you're able to resolve that problem, the idea of the passthrough mode feature/patch might not make sense.

That said to debug further about what's going on with the power axis, you could try pulling this new version with debug logging here:

https://github.com/joestringer/pcsx2/actions

You might need to remap the controls again, I'm not sure. But after that you can go to `Tools` -> `Enable Log Window` or `Enable File Logging`. When you run the game, the log will spam lines about the brake and power values. If you can start with the controller with brakes and power off, then move the power lever from released up to full power and back down, then scan up through the logs until you see logs printing different values for the power axis and share those logs, it would help to understand the range of values that are being pulled into this code. Separately if you can figure out the brake, then it would also help to have the same log data lines with the varying brake input values.

sonik-br commented:
@joestringer tested...
Forcing it on the ini file to fullaxis as:
```
TrainController_Power = DInput-0/FullAxis1
TrainController_Brake = DInput-0/FullAxis0
```

Brake only shows as FF on it's first position and FF on all other positions.
Power shows as:

P5 80
P4 A0
P3 Be
P2 D3
P1 ED
N FF
Transiton FF

All tests doe using direct input.

And interesting: Same results with and without passthrough mode

joestringer commented:
> And interesting: Same results with and without passthrough mode

Oh yeah maybe I didn't explain well. The logs you see are presenting the raw input from DInput, before this code processes the values to output to the game. So the passthrough mode doesn't have an effect on the values you see there. That's normal.

I can't really explain the discrepancy.. If you're running a DGOC-44U, then [these docs](https://marcriera.github.io/ddgo-controller-docs/controllers/usb/dgoc44u/) describe the notches as:

N | P1 | P2 | P3 | P4 | P5 | Transition
-- | -- | -- | -- | -- | -- | --
0x81 | 0x6D | 0x54 | 0x3F | 0x21 | 0x00 | 0xFF

Comparing this with the values that you have presented, there values are very different:

N | P1 | P2 | P3 | P4 | P5 | Transition
-- | -- | -- | -- | -- | -- | --
0xFF | 0xED | 0xD3 | 0xBe | 0xA0 | 0x80 | 0xFF

The passthrough mode was written with the idea in mind that the values from the input library would be identical to the selected subtype of controller. But even if I look through the other controller types in the Marc Riera datasheets, I can't find any controller types that match the values you are reporting.

One interesting thing about the bitpatterns you report is that the bit at offset 0x80 is always set. On average across the values they're all around 0x7E-0x80 higher than the value reported by the docs. This makes me wonder whether either model you have works differently somehow, or if perhaps the values get treated as signed values that get set negative somewhere, or if somehow the input library is doing some conversions that cause a skewing of the values into the upper range.. Not sure if `FullAxis` vs. `+Axis` vs `-Axis` configuration could have this sort of impact as well.

Technically it's not that difficult to try to adjust for this, but it's a bit hard for me to picture the range of devices that may be impacted by this behaviour - is this a general property of all of the original controllers so I should apply a filter automatically in passthrough mode? Or is it that the axis needs to be configured in a particular way to cause these results? Will a fix I come up with help or hurt the emulation of this controller? I'd welcome input from upstream PCSX2 devs if they have ideas on how to best handle this. I guess I could add another separate setting option called "DGOC-44U compatibility" if we think that it's this specific controller that has this behaviour. On the other hand, another option is just to remove passthrough mode if it's not going to fully work for someone.

In the mean time, just to see if it could work, I added a hack to the latest builds over here to try to have the right behaviour for your environment: https://github.com/joestringer/pcsx2/actions .

> Brake only shows as FF on it's first position and FF on all other positions.

OK. That confirms that Dinput isn't processing the brake signal at all. Pretty much what we expected since you can't even configure it through the UI.

sonik-br commented:
@joestringer I'm not exactly using a real "DGOC-44U"...
It's a PS2 Type 2 device with a custom made usb adapter that outputs like a DGOC-44U.
But I don't think that's the problem as the PC version of the game detects and uses the device correctly.

Btw looks like the power axis is working as half axis, right? It goes from 0x80 to 0xff.
It's like if dinput apply some kind of range calibration.

Florin9doi commented:
Try to switch from dinput to sdl

sonik-br commented:
Testing under windows joy.cpl

Neutral
![image](https://github.com/user-attachments/assets/fa8ae9bc-2317-4243-a599-d32dd9f342f1)


P5
![image](https://github.com/user-attachments/assets/62028698-2ba9-45b9-88fb-2d5f0a7ad179)

joestringer commented:
Yeah if you can try SDL, that would be good. But I would not suggest using the latest build I shared from my GitHub fork, maybe go back to the previous one that you tested or just pick up the build from this PR. The latest build on my fork will mess with the input as it gets passed through.

joestringer commented:
joy.cpl output there looks good, matches the spec I expected.

sonik-br commented:
SDL raw, with the same build as before
P5 ff
P4 be
P3 82
P2 57
P1 24
N 00
Transiton 00

joestringer commented:
Those latest numbers look more like the scale is inverted and spread across the full 256-bit range rather than limiting to 128 bits. That I could see being part of how the input is processed and rendered into floats before handing to this patch. 

Is that configuration with a half axis or a full axis configured? Also, does brake work with SDL or do you see the same issue with the brakes?

mario032106 commented:
Shinkansen USB passthrough isn't working - it says both handles have a value of 0, and actually moving the physical controller does nothing. Especially annoying, as "removing" the virtual USB freezes the game.

joestringer commented:
@mario032106 thanks for trying it out! Would you be able to also share the console logs using these builds https://github.com/joestringer/pcsx2/actions/runs/10501516994 , similar to the information I asked for in [this comment](https://github.com/PCSX2/pcsx2/pull/11723#issuecomment-2303623822)? There could be something fairly simple that I got wrong since I don't have the hardware to test with.

mario032106 commented:
@joestringer Used that exact build, and had it log to console. Both handles started at 0. No input was registered, not even the face buttons. Mapped it to the keyboard to check if it even worked, and it did, so either it's a problem with my controller (unlikely, as AutoTrainTAS recognizes it just fine, as does Final on an actual PS2) or USB passthrough.

joestringer commented:
@mario032106 and to confirm is that using SDL for the input? The logs print the raw values pulled from the input library, so if those are printing zeroes regardless of the input then it's before this code even gets reached.

Florin9doi commented:
> Shinkansen controller USB passthrough isn't working - it says both handles have a value of 0, and actually moving the physical controller does nothing. Especially annoying, as "removing" the virtual USB freezes the game.

There is no USB passthrough , only axis passthrough. A real Shinkansen wont work unless you have a 3rd party HID driver

sonik-br commented:
> Is that configuration with a half axis or a full axis configured? Also, does brake work with SDL or do you see the same issue with the brakes?

Test was done using full axis.

```
[USB1]
Type = TrainController
TrainController_Passthrough = true
TrainController_Power = SDL-0/FullAxis7
TrainController_Brake = SDL-0/FullAxis6
```

For the Brake, same as before.
Brake only shows as 0xF9 on it's first position and 0xFF on all other positions.

My previous post was wrong, it shows 0xF9 on the first position. Sorry.

mario032106 commented:
> @mario032106 and to confirm is that using SDL for the input? 
I didn't even know it was supposed to show up as an SDL input device. I assumed USB passthrough fed the raw data to the game, and vice-versa (for the tiny display). I used Zadig to install the WinUSB driver for it so I could test it with AutoTrainTAS (couldn't get it to recognize it otherwise), could it be that that's interfering with it?

joestringer commented:
I may have underestimated the complexity of connecting this logic with real-world DDG controllers ;) Sorry about that. This PR was initially designed just so I could get my Zuiki controller to work. The Zuiki connects like a regular joystick controller into Windows, then outputs various axis values through the positive and negative Y axes when I connect it using SDL. With the passthrough mode turned off, that maps accurately into the game and I can play without any issues. I assume that any other joystick which outputs values in a similar manner should work the same, potentially including things like HOTAS controllers (though I also don't have one of those to try out).

What I didn't realize is that some of these older joysticks do not implement HID, and that there may be issues with some of the input libraries in getting the signal into the code that this PR touches, whether SDL, Xinput or Dinput. When I sketched up the patch for axis passthrough, I knew that the baseline functionality made different assumptions about the joystick's output value range compared with the [spec sheets](https://marcriera.github.io/ddgo-controller-docs/), so in an ideal world I would take whatever signal that those controllers provided on the relevant axis and send it straight to the game. My naive _hope_ was that this would "just work", but I don't have the physical hardware to test & debug it. As y'all have provided feedback, it's becoming more and more obvious that there are subtleties to the problem space that I just don't know about. I don't necessarily know enough at the moment to be able to solve all the problems for everyone. 

For @sonik-br's report, I could potentially get it working by tweaking the options to allow (a) inverting the axis and (b) scaling the full 256-value axis down into a 128-bit output value to the game. I can throw that together. That said, if the brake is still not pushing a full range of values into the controller emulation layer through SDL then I can't solve that. In the end you're gonna want to use both power and brake to actually play.

Similarly, for @mario032106, if you can get both axes configured in the settings window for power/brake (and all the buttons), and if that then outputs different values into the console, then there's a chance I could introduce tweaks into these patches to match the input to support your joystick. There may be a bit of tedious back/forth to get it working, but it should be doable as long as the pcsx2 controller emulation layer can get the full range of axis signal inputs.

Regardless of all that, even without the USB passthrough support, I think this PR could be useful at least for players with a Zuiki, if not some other lever-like joystick controllers, so worst case I could just drop the USB passthrough patch and defer that to a subsequent PR (and maybe try to work more closely with some of you to get this emulation working with a wider range of hardware controllers).

joestringer commented:
Oh, and naming is one of the hardest problems in computer science ;) Maybe it would help if I came up with a better name than "USB passthrough" as I agree with @mario032106 that the name is a bit misleading since it can't get raw access to the original controllers.

joestringer commented:
> I used Zadig to install the WinUSB driver for it so I could test it with AutoTrainTAS (couldn't get it to recognize it otherwise), could it be that that's interfering with it?

As far as I understand, the denconv tool by AutoTrainTAS takes input from Windows input and then monitors for the DDG programs running and edits the memory of the running application to inject the controls into the game. Assuming that's correct, I wouldn't expect that denconv interferes with this. I could always be wrong so let me know if you observe otherwise.

sonik-br commented:
@joestringer thanks for taking your time and doing this!

In my case, I'm using a (open source and simple to build) custom usb adapter to read a PS2 device (non-hid) and output as a PC One handle or Two handle device.
I can also make it output like the ddgo Nintendo Switch controller, or even create a custom output, to match what the emulator (and your code) expects.

Florin9doi commented on pcsx2/USB/usb-pad/usb-train.cpp:
```diff
@@ -0,0 +1,555 @@
+
+		usb_desc_init(&s->dev);
+		usb_ep_init(&s->dev);
+		train_handle_reset(&s->dev);
```
The passthrough value isn't read in CreateDevice like this : 
https://github.com/PCSX2/pcsx2/blob/fd5a6522702c3279c2a968f395983a46374b7fa2/pcsx2/USB/usb-lightgun/guncon2.cpp#L469
This is why passthrough axes are messed up.

Florin9doi replied:
This should fix DGOC-44U

joestringer replied:
Oh, that was an oversight on my part. Nice find, will fix this up.

Florin9doi commented on pcsx2/USB/usb-pad/usb-train.cpp:
```diff
@@ -0,0 +1,555 @@
+		static constexpr const SettingInfo passthrough = {
+			SettingInfo::Type::Boolean,
+			"Passthrough",
+			TRANSLATE_NOOP("USB", "USB Passthrough Mode"),
+			TRANSLATE_NOOP("USB", "Passes through the input axis to the game. Enable if you are using an original Densha De Go! controller. Disable if you are using any other joystick."),
```
```suggestion
			TRANSLATE_NOOP("USB", "Axes Passthrough"),
			TRANSLATE_NOOP("USB", "Passes through the input axis unprocessed to the game. Enable if you are using a compatible Densha De Go! controller. Disable if you are using any other joystick."),
```

joestringer replied:
Applied, thanks.

joestringer commented:
@Florin9doi ah, the extra pair of eyes is much welcomed. I fixed that up and confirmed with the latest version that the unprocessed input values are passed through to the game.

@sonik-br very cool! Glad to hear the Nintendo DDG mode is working for you. Hopefully now the axis passthrough mode should also work for DGOC-44U. It still won't resolve any issues if the input library can't detect the axis, but it should fix issues where PCSX2 reads the axis value and interprets it in a way that causes the throttle or brake to be "stuck" at some level and won't change when you move the axis.

I've updated the branch here with the latest fixes and without the verbose logging to the console. I've also pushed a separate build [here](https://github.com/joestringer/pcsx2/pull/2) with additional logging in case anyone wants to try to figure out what's happening with the raw inputs. Builds available here: https://github.com/joestringer/pcsx2/actions .

sonik-br commented:
Still a no go for me with DGOC-44U.
Same range as before (0x80 to 0xff)

Power 5 shows as:
DenshaCon data: brake={in:00,out:00} power={in:80,out:80} type=0 passthrough=1

joestringer commented:
@sonik-br OK, I've just pushed a new experimental approach to this over at https://github.com/joestringer/pcsx2/pull/2 / https://github.com/joestringer/pcsx2/actions. With the new approach there is a specific setting for DGOC-44U which will adjust the power values based on the input you've provided earlier in this thread. You could give that a go and see if it helps.

sonik-br commented:
Tested.
I believe that the out values are off...

```
Type = TrainController
TrainController_Compatibility = DGOC-44U
TrainController_Power = SDL-0/FullAxis7
```

```
N DenshaCon data: brake={in:00,out:79} power={in:ff,out:7f} type=0 compat=2
1 DenshaCon data: brake={in:00,out:79} power={in:ed,out:6d} type=0 compat=2
2 DenshaCon data: brake={in:00,out:79} power={in:d3,out:53} type=0 compat=2
3 DenshaCon data: brake={in:00,out:79} power={in:be,out:3e} type=0 compat=2
4 DenshaCon data: brake={in:00,out:79} power={in:a0,out:20} type=0 compat=2
5 DenshaCon data: brake={in:00,out:79} power={in:80,out:00} type=0 compat=2
```

joestringer commented:
👍 I see. Does the game recognize the notches or not? It's weird to me that notch 1 and 5 are correct but the others are off by a little. I can hack this to make it work accurate for this case, but I don't know why this is happening and whether it would help for others running the same controller. Either way I uploaded new builds at the same locations that should capture the right values hopefully. Not sure how the behaviour will be when the lever is in transition though, since the Neutral value is the same as transition in your reports.

sonik-br commented:
This fixed the power handle!
But brake still shows as input 0xf9 and 0xff only.
I believe that this might be related to windows only. Windows does not pass raw HID data.
Not sure if there's anything that can be done to fix this.

Florin9doi commented:
I made a dgoc44u clone and I found that both windows and linux are reading the axes as signed int8 although the controller [declares](https://github.com/marcriera/ddgo-controller-docs/blob/b80725296609a76f77333d96d7fe4e82f148c776/content/controllers/usb/descriptors/dgoc44u_hid-report-descriptor.txt#L8-L9
) positive values and so half of values are reported as negative. And even more, Windows applies some built-in calibration and the values get shifted by one near the center.

This is a draft for dgoc44u, but it should be probably applied behind a new setting; signed/unsigned mode. 0.4995f is used to compensate for calibration/rounding error, but would be better to stick to the nearest known value.

```diff
               switch (bind_index)
                {
                        case CID_TC_POWER:
-                               s->data.power = static_cast<u32>(std::clamp<long>(std::lroundf(value * 255.0f), 0, 255));
+                               if (value < 0.5f)
+                                       s->data.power = 0x81 + static_cast<u32>(std::clamp<long>(std::lroundf(value * 255.0f), 0, 255));
+                               else
+                                       s->data.power = static_cast<u32>(std::clamp<long>(std::lroundf((value - 0.4995f) * 255.0f), 0, 255));
                                break;
                        case CID_TC_BRAKE:
-                               s->data.brake = static_cast<u32>(std::clamp<long>(std::lroundf(value * 255.0f), 0, 255));
+                               if (value < 0.5f)
+                                       s->data.brake = 0x81 + static_cast<u32>(std::clamp<long>(std::lroundf(value * 255.0f), 0, 255));
+                               else
+                                       s->data.brake = static_cast<u32>(std::clamp<long>(std::lroundf((value - 0.5f) * 255.0f), 0, 255));
                                break;
```

lightningterror commented:
Is this good to go for merge?

Florin9doi commented:
> Is this good to go for merge?

Yes, one of the features doesn't work as intended, but it can be improved later.

joestringer commented:
Ah right, good spotting @Florin9doi.. I wasn't taking into account that the raw values are floats and by the time my debug logging printed the values, the values were already excluding values outside the input float values of 0.0-1.0. 

I pushed a fresh build with this latest heuristic and a new compatibility mode setting here: https://github.com/joestringer/pcsx2/pull/2 . If anyone wants to try that out, make sure to set the DGOC-44U compatibility mode in the controller settings.

sonik-br commented:
Still not working for me :/

Brake only shows as:
brake_in=1.000000 brake_out=128 type=0 compat=1
brake_in=0.976195 brake_out=122 type=0 compat=1

And Power fron none to 5 (including transition state that shows as 1.0) :

power_in=1.000000 power_out=128 type=0 compat=1
power_in=0.928571 power_out=109 type=0 compat=1
power_in=1.000000 power_out=128 type=0 compat=1
power_in=0.829370 power_out=84 type=0 compat=1
power_in=1.000000 power_out=128 type=0 compat=1
power_in=0.746040 power_out=63 type=0 compat=1
power_in=1.000000 power_out=128 type=0 compat=1
power_in=0.627002 power_out=33 type=0 compat=1
power_in=1.000000 power_out=128 type=0 compat=1
power_in=0.500000 power_out=0 type=0 compat=1

joestringer commented:
OK, the power_in values seem to be a bit different from what @Florin9doi was reporting with their simulated controller. Again though even accounting for the raw input from the library I can't explain the limitations in the brake inputs.. that would have to be fixed separately to get this emulation feature working with that controller.

shartcarbuncle commented:
First of all, thanks for adding train controller support to PCSX2. I've had a blast over the past year using the Zuiki with *Final*, *Professional 2* and *Ryojohen*.

I did find a problem, however, with *Densha de Go 3*. Because we're impersonating a two handle controller, there's no good way to select the time of day when starting a route. When using a two handle controller, you choose the route with your power lever, and then change the time of day with the brake lever. Since both are tied together on the single Zuiki lever, you can't change them independently.

One workaround I've found for this is to use a regular gamepad to make the selection, then basically hot swap to the train controller; this sorta works, but the UI naturally doesn't reflect the change, and you can only do it once before you have to reset the game.

I think the best solution would be to add a profile for the one handle controller. https://marcriera.github.io/ddgo-controller-docs/controllers/classic/tcpp20001/

It's natively supported by DDG 3, and allows for using the face buttons for the time of day setting.

I know it's been an entire year, so I totally understand if you've moved on with your life and don't want to touch this stuff again, but I thought it couldn't hurt to post my findings. If I get brave, I may try poking around myself and breaking everything.

So far, all I've managed is a a first draft vector art representation of the controller.


<img width="5261" height="2980" alt="one hand mascon vector" src="https://github.com/user-attachments/assets/29b069f4-e29c-48cd-baa4-279194c063bd" />
